### ISSUE TITLE
Error in hostcalls functions causes filter to panic

### ISSUE THREAD

**Issue by alandiegosantos** [proxy-wasm/proxy-wasm-rust-sdk#82]
## Summary
The hostcalls panic if the status is not expected. It causes filter to stop working if there is any non-expected status response in any ABI call.
```
[2021-02-24 23:03:59.403][25254][critical][wasm] [external/envoy/source/extensions/common/wasm/context.cc:1157] wasm log: panicked at 'unexpected status: 2', external/raze__proxy_wasm__0_1_3/src/hostcalls.rs:233:23
```

## Expectation
The Status error is propagated through the context calls, so each dev can deal with the error.

**Comment by alandiegosantos** [comment]
#83

**Comment by PiotrSikora** [comment]
How are you getting this panic and what proxy (and version) are you using? Do you have a simple reproducer?

If both proxy and SDK implement the same ABI correctly, then there are no unexpected status codes, and you should never see that panic, but clearly there is something off here.

Note that _expected_ errors are propagated through context call.

**Comment by alandiegosantos** [comment]
The panic was observed on Envoy 1.17 on OSX when the client disconnects while waiting for the response.
```
$ envoy --version
envoy  version: 5c801b25cae04f06bf48248c90e87d623d7a6283/1.17.0/Modified/DEBUG/BoringSSL
```
And, this is the *on_log* fn:
```
impl HttpContext for PanicFilter {
    fn on_log(&mut self) {
        if self.get_http_response_header(":status").is_none() {
            error!("Client disconnected while waiting for the response");
        }
    }
}
```
And these are the generated logs.
```
[2021-02-27 00:08:58.637][62818][critical][wasm] [source/extensions/common/wasm/context.cc:1180] wasm log: panicked at 'unexpected status: 2', external/raze__proxy_wasm__0_1_3/src/hostcalls.rs:233:23
[2021-02-27 00:08:58.640][62818][error][wasm] [source/extensions/common/wasm/wasm_vm.cc:39] Function: proxy_on_log failed: Uncaught RuntimeError: unreachable
Proxy-Wasm plugin in-VM backtrace:
  0:  0x11446 - rust_panic
  1:  0x111e8 - _ZN3std9panicking20rust_panic_with_hook17h4f753dc70b771d8eE
  2:  0x11680 - _ZN3std9panicking19begin_panic_handler28_$u7b$$u7b$closure$u7d$$u7d$17h5c9fc29957309ae8E
  3:  0x115c7 - _ZN3std10sys_common9backtrace26__rust_end_short_backtrace17h92808941d97e9b21E
  4:  0x61de - rust_begin_unwind
  5:  0xc48 - _ZN3std9panicking15begin_panic_fmt17hf0ddb2cbe0cdc73aE
  6:  0x13ad - _ZN10proxy_wasm9hostcalls13get_map_value17ha42493ab0e1679c1E
  7:  0x4f43 - _ZN77_$LT$client_stats..HttpMetrics$u20$as$u20$proxy_wasm..traits..HttpContext$GT$6on_log17haa2cc2e952d59848E
  8:  0xeee9 - proxy_on_log
```

If I replace the *self.get_http_response_header(":status")* with *self.get_property(vec!["response", "headers", ":status"])*, the filter does not panic. I expected to receive an empty Option if the header is not present. This filter works fine if the response arrived.

The fn *self.get_http_response_headers()* works and returns an empty *Vec<String, String>*. We could patch the function *self.get_http_response_header("header")* to use *get_property* or look for the header in the *Vec<String, String>* from  *get_http_response_headers*.

**Comment by PiotrSikora** [comment]
Thanks for the reproducer! This is actually an bug in the Envoy implementation, see: https://github.com/envoyproxy/envoy/pull/15224.

**Comment by PiotrSikora** [comment]
Fixed in envoyproxy/envoy#15224.
